### ISSUE TITLE
using sampled softmax is slow in 'double-tower' like model

### ISSUE THREAD

**Issue by jaydone78** [tensorflow/tensorflow#49145]
I'm trying to implement a 'double tower' like recommendation model, as something described in paper 'Sampling-Bias-Corrected Neural Modeling for Large Corpus Item Recommendations'. It mentioned to use 'in batch softmax' to learn the embedding. I think this 'in batch softmax' is very likely to the 'sampled softmax' used in word2vec. So I tried to use the tf.nn.nce_loss to implement this logic. I wrote some code try to train a simple model.

    
    graph = tf.Graph()
    vocabulary_size = 300000
    embedding_size = 128
    dense_size = 64
    batch_size = 256
    
    import math
    
    with graph.as_default():
    # Input data.

    train_inputs = tf.placeholder(tf.int32, shape=[batch_size])
    train_labels = tf.placeholder(tf.int32, shape=[batch_size, 1])
    

    # Ops and variables pinned to the CPU because of missing GPU implementation
    with tf.device('/cpu:0'):
        # Look up embeddings for inputs.
        embeddings = tf.Variable(
            tf.random_uniform([vocabulary_size, embedding_size], -1.0, 1.0))
        user = tf.nn.embedding_lookup(embeddings, train_inputs)
        user = tf.layers.dense(user, dense_size, activation='relu')
        user = tf.nn.l2_normalize(user, axis=-1)
        print(user)
        
        right_embedding = tf.Variable(
            tf.random_uniform([vocabulary_size, embedding_size], -1.0, 1.0))
        right_hidden = tf.layers.dense(right_embedding, dense_size, activation='relu')
        
        
        nce_biases = tf.Variable(tf.zeros([vocabulary_size]),dtype=tf.float32,trainable=False)

        loss = tf.reduce_mean(
            tf.nn.nce_loss(weights=right_hidden,biases=nce_biases, inputs=user, labels=train_labels,
                 num_sampled=num_sampled, num_classes=vocabulary_size))
        
        optimizer = tf.train.GradientDescentOptimizer(1.0).minimize(loss)
        init = tf.global_variables_initializer()

and I found that the training is really slow. I doubt that it's because every time tensorflow need to calculate the right_hidden = tf.layers.dense(right_embedding, dense_size, activation='relu'), and the sampled softmax only applies to the last matrix "right_hidden", but we still need to calculate the entire right_hidden matrix beforehand. So am I right? or anyone can give a better solution to deal with the 'sampled softmax' or 'in batch softmax'. thanks

**Comment by tilakrayal** [comment]
@jaydone78 ,

Please, fill [issue template](https://github.com/tensorflow/tensorflow/issues/new/choose).

Please, share colab link or simple standalone code with supporting files to reproduce the issue in our environment.It helps us in localizing the issue faster.Thanks!

**Comment by jaydone78** [comment]
I've upload the training data and code in the colab: 
https://colab.research.google.com/drive/11prB0qNvxXAiE-zzD9TtBn4OS8ozVeY_?usp=sharing
